pkg/distributor/distributor_ingest_storage_test.go

```diff
@@ -260,6 +264,306 @@ func TestDistributor_Push_ShouldSupportIngestStorage(t *testing.T) {
+					makeWriteRequest(0, 1, 0, false, false, "series_4", "series_5"),
+				},
+				"zone-b": {
+					nil,
+					nil,
+				},
+			},
+			expectedSeries: 5,
+		},
+		"partitions RF=2 (2 zones), 4 ingesters, ingesters owning different partitions are UNHEALTHY across both zones": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {states: []ingesterState{ingesterStateFailed, ingesterStateHappy}},
+				"zone-b": {states: []ingesterState{ingesterStateHappy, ingesterStateFailed}},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					nil,
+					makeWriteRequest(0, 1, 0, false, false, "series_4", "series_5"),
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "series_1", "series_2", "series_3"),
+					nil,
+				},
+			},
+			expectedSeries: 5,
+		},
+		"partitions RF=2 (2 zones), 4 ingesters, ingesters owning the same partition are UNHEALTHY in both zones": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {states: []ingesterState{ingesterStateHappy, ingesterStateFailed}},
+				"zone-b": {states: []ingesterState{ingesterStateHappy, ingesterStateFailed}},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					makeWriteRequest(0, 1, 0, false, false, "series_1", "series_2", "series_3"),
+					nil,
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "series_1", "series_2", "series_3"),
+					nil,
+				},
+			},
+			expectedErr: errFail,
+		},
+		"partitions RF=2 (2 zones), 4 ingesters, ingesters owning the same partition are UNHEALTHY in both zones but the partition is not part of the tenant's shard": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {states: []ingesterState{ingesterStateFailed, ingesterStateHappy}},
+				"zone-b": {states: []ingesterState{ingesterStateFailed, ingesterStateHappy}},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					nil,
+					makeWriteRequest(0, 1, 0, false, false, "series_1", "series_2", "series_3"),
+				},
+				"zone-b": {
+					nil,
+					makeWriteRequest(0, 1, 0, false, false, "series_1", "series_2", "series_3"),
+				},
+			},
+			shardSize:      1, // Tenant's shard made of: partition 1.
+			expectedSeries: 3,
+		},
+	}
+
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			for _, minimizeIngesterRequests := range []bool{false, true} {
+				minimizeIngesterRequests := minimizeIngesterRequests
+
+				t.Run(fmt.Sprintf("minimize ingester requests: %t", minimizeIngesterRequests), func(t *testing.T) {
+					t.Parallel()
+
+					// Find the max number of ingesters in a zone. It will be used as the number of partitions.
+					numPartitions := 0
+					for _, state := range testData.ingesterStateByZone {
+						numPartitions = max(numPartitions, max(state.numIngesters, len(state.states)))
+					}
```

Comment:
Suggestion is to add a function on prepConfig to set a field on prepConfig itself. Don't think will make code better.

Reply:
the suggestion was to add `maxIngestersPerZone() int` function and use that here. Mainly because `prepConfig` already counts ingesters and zones and know what are the different ways of setting the number of ingesters (`states` vs happy/unhappy)

it's a minor change, i'll leave it up to you

Reply:
Instead of explicitly setting `ingestStoragePartitions` we can auto-detect it in most cases. I've done it here: https://github.com/grafana/mimir/pull/7402/commits/1b44a0eabb45adcf4f93c3c5ecbe9faa779f4f19